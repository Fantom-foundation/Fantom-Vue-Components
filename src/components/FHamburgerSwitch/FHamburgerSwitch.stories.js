import FHamburgerSwitch from './FHamburgerSwitch.vue';

export default {
    title: 'FHamburgerSwitch',
    component: FHamburgerSwitch,
};

export const Default = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <f-hamburger-switch />
        </div>
    `,
});

export const Thickness = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <f-hamburger-switch thickness="1" />
            <f-hamburger-switch thickness="2" />
            <f-hamburger-switch thickness="3" />
        </div>
    `,
});

export const TwoLines = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <f-hamburger-switch two-lines thickness="1" />
            <f-hamburger-switch two-lines thickness="2" />
            <f-hamburger-switch two-lines thickness="3" />
        </div>
    `,
});

export const MobileViewBreakpoint = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <h3>Resize viewport under/above 800px</h3>
            <f-hamburger-switch mobile-view-breakpoint="800px" :on="on" thickness="2" />
        </div>
    `,
    data() {
        return {
            on: true,
        };
    },
    mounted() {
        setTimeout(() => {
            this.on = false;

            setTimeout(() => {
                this.on = true;
            }, 100);
        }, 100);
    },
});

export const Size = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <f-hamburger-switch :style="bigSize" />
            <f-hamburger-switch :style="smallSize" />
        </div>
    `,
    data() {
        return {
            bigSize: {
                transform: 'scale(1.2, 1.2)',
            },
            smallSize: {
                transform: 'scale(0.5, 0.5)',
            },
        };
    },
});

export const Colors = () => ({
    components: { FHamburgerSwitch },
    template: `
        <div>
            <f-hamburger-switch :style="blueStyle" thickness="2" />
            <f-hamburger-switch :style="purpleStyle" thickness="2" />
        </div>
    `,
    data() {
        return {
            blueStyle: {
                '--fhamburgerswitch-color': '#08209e',
            },
            purpleStyle: {
                '--fhamburgerswitch-color': '#800E99',
            },
        };
    },
});
