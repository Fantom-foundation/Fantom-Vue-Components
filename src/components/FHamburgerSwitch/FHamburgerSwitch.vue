<template>
    <div
        class="fhamburgerswitch"
        role="switch"
        :class="[classes, `fhamburgerswitch-thickness${thickness}`]"
        @click="toggle"
    >
        <div class="cont">
            <div class="l1"></div>
            <div class="l2" v-if="!twoLines"></div>
            <div class="l3"></div>
        </div>
    </div>
</template>

<script>
import { breakpointsMixin } from '../../mixins/breakpoints.js';
import Vue from 'vue';

/** Global state for FHamburgerSwitch(es) */
const switchState = new Vue({
    data: {
        on: false,
    },
});

/**
 * Menu switch with animation -> 3 (2) horizontal lines to `X`.
 * It is controlled by global state, so if one instance of `FHamburgerSwitch` is turned on, the other instances are also turned on.
 */
export default {
    name: 'FHamburgerSwitch',

    mixins: [breakpointsMixin],

    props: {
        /** Turn the switch on on initialization */
        on: {
            type: Boolean,
            default: false,
        },
        /** Use just 2 lines instead of 3. */
        twoLines: {
            type: Boolean,
            default: false,
        },
        /** Thickness of one switch row. */
        thickness: {
            type: String,
            default: '1',
            validator: function(_value) {
                return ['1', '2', '3'].indexOf(_value) !== -1;
            },
        },
        /** Minimum width for switching to mobile view */
        mobileViewBreakpoint: {
            type: String,
            default: '',
        },
    },

    computed: {
        isOn() {
            return switchState.on;
        },

        classes() {
            return {
                'fhamburgerswitch-on': this.isOn,
            };
        },
    },

    watch: {
        on(value) {
            switchState.on = !!value;
        },

        isOn(value) {
            const eBody = document.body;

            if (value) {
                eBody.classList.add('fhamburgerswitch-on');
            } else {
                eBody.classList.remove('fhamburgerswitch-on');
            }
        },
    },

    created() {
        switchState.on = this.on;

        if (this.mobileViewBreakpoint) {
            this.breakpoints = [{ value: this.mobileViewBreakpoint }];
        }
    },

    methods: {
        toggle() {
            switchState.on = !switchState.on;

            if (switchState.on) {
                this.$emit('hamburger-switch-on');
            } else {
                this.$emit('hamburger-switch-off');
            }
        },

        onBreakpointChange() {
            switchState.on = false;
        },
    },

    on() {
        switchState.on = true;
    },

    off() {
        switchState.on = false;
    },
};
</script>

<style lang="scss">
@use "style";
</style>
