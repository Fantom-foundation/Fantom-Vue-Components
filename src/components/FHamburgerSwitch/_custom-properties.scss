.fhamburgerswitch {
    --fhamburgerswitch-color: var(--f-color-primary-5, #1a6aff);
    --fhamburgerswitch-transition-length: var(--f-transition-length, 220ms);
    --fhamburgerswitch-transition-func: var(--f-transition-func, ease);
}
