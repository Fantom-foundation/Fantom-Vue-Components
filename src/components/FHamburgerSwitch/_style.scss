@use "../../assets/scss/tools/media";
@use "custom-properties";

.fhamburgerswitch {
    display: inline-block;
    width: 24px;
    height: 24px;
    cursor: pointer;
    /*transform: scale(1.1);*/

    .cont {
        position: relative;
        width: 100%;
        height: 100%;
        /*border: 1px solid transparent;*/
    }

    .l1,
    .l2,
    .l3 {
        position: absolute;
        width: 100%;
        height: 2px;
        background-color: var(--fhamburgerswitch-color);
        border-radius: 3px;
        transition: all var(--fhamburgerswitch-transition-length) var(--fhamburgerswitch-transition-func);
    }

    .l1 {
        top: 0;
    }

    .l2 {
        top: 50%;
    }

    .l3 {
        bottom: 0;
    }

    &-thickness1 {
        .l1 {
            top: 10%;
        }

        .l2 {
            margin-top: -1px;
        }

        .l3 {
            bottom: 10%;
        }
    }

    &-thickness2 {
        .l1,
        .l2,
        .l3 {
            height: 4px;
        }

        .l1 {
            top: 5%;
        }

        .l2 {
            margin-top: -2px;
        }

        .l3 {
            bottom: 5%;
        }
    }

    &-thickness3 {
        .l1,
        .l2,
        .l3 {
            height: 6px;
        }

        .l2 {
            margin-top: -3px;
        }
    }

    &-on {
        .l1 {
            /*opacity: 1;*/
            /*transform: translate(0, 10px) rotate(45deg);*/
            top: 0;
            transform: translate(0, 11px) rotate(45deg);
        }

        .l2 {
            opacity: 0;
        }

        .l3 {
            /*opacity: 1;*/
            bottom: 0;
            transform: translate(0, -11px) rotate(-45deg);
        }

        &.fhamburgerswitch-thickness2 {
            .l1 {
                transform: translate(0, 10px) rotate(45deg);
            }

            .l3 {
                transform: translate(0, -10px) rotate(-45deg);
            }
        }

        &.fhamburgerswitch-thickness3 {
            .l1 {
                transform: translate(0, 9px) rotate(45deg);
            }

            .l3 {
                transform: translate(0, -9px) rotate(-45deg);
            }
        }
    }
}


@include media.prefers-reduced-motion() {
    .fhamburgerswitch {
        .l1,
        .l2,
        .l3 {
            transition: unset;
        }
    }
}
