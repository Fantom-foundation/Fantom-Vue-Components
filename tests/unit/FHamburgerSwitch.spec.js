import { mount } from '@vue/test-utils';
import FHamburgerSwitch from '@/components/FHamburgerSwitch/FHamburgerSwitch.vue';

let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
    }
}

function createWrapper({ propsData = {}, attachTo } = {}) {
    return mount(FHamburgerSwitch, {
        propsData,
        attachTo,
    });
}

beforeEach(() => {
    wrapper = createWrapper();
});

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FHamburgerSwitch', () => {
    it('should render correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should have `on` prop default to `false`', () => {
        expect(wrapper.props().on).toBe(false);
    });

    it('should have `twoLines` prop default to `false`', () => {
        expect(wrapper.props().twoLines).toBe(false);
    });

    it("should have `mobileViewBreakpoint` prop default to `''`", () => {
        expect(wrapper.props().mobileViewBreakpoint).toBe('');
    });

    it("should have `thickness` prop default to '1'", () => {
        expect(wrapper.props().thickness).toBe('1');
    });

    it('should turn the switch on if `on` prop is set to `true`', () => {
        destroyWrapper(wrapper);
        wrapper = createWrapper({ propsData: { on: true } });

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(true);
    });

    it('should add correct `fhamburgerswitch-thickness` css class if `thickness` prop is set', () => {
        // default
        expect(wrapper.classes('fhamburgerswitch-thickness1')).toBe(true);

        destroyWrapper(wrapper);
        wrapper = createWrapper({ propsData: { thickness: '2' } });

        expect(wrapper.classes('fhamburgerswitch-thickness2')).toBe(true);
    });

    it('should toggle the switch by changing `on` prop', async () => {
        await wrapper.setProps({ on: true });

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(true);

        await wrapper.setProps({ on: false });

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(false);
    });

    it('should toggle the switch by clicking on it', async () => {
        expect(wrapper.classes('fhamburgerswitch-on')).toBe(false);

        await wrapper.trigger('click');

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(true);
    });

    it('should turn the switch on by calling static method `FHamburgerSwitch.on()`', async () => {
        FHamburgerSwitch.on();

        await wrapper.vm.$nextTick();

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(true);
    });

    it('should turn the switch off by calling static method `FHamburgerSwitch.off()`', async () => {
        await wrapper.trigger('click');

        FHamburgerSwitch.off();

        await wrapper.vm.$nextTick();

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(false);
    });

    it('should turn the switch off if `mobileViewBreakpoint` is set and reached', async () => {
        /*destroyWrapper(wrapper);
        wrapper = createWrapper({
            propsData: { mobileViewBreakpoint: '40em' },
            attachTo: document.body,
        });*/

        await wrapper.trigger('click');

        wrapper.vm.onBreakpointChange();

        await wrapper.vm.$nextTick();

        expect(wrapper.classes('fhamburgerswitch-on')).toBe(false);
    });

    it('should emit `hamburger-switch-on` and `hamburger-switch-off` events by toggling the switch', async () => {
        await wrapper.trigger('click');
        await wrapper.trigger('click');

        expect(wrapper.emitted()['hamburger-switch-on']).toBeTruthy();
        expect(wrapper.emitted()['hamburger-switch-on'].length).toBe(1);
        expect(wrapper.emitted()['hamburger-switch-off']).toBeTruthy();
        expect(wrapper.emitted()['hamburger-switch-off'].length).toBe(1);
    });

    it('should add/remove css class `fhamburgerswitch-on` to/from the `body` element by toggling the switch', async () => {
        await wrapper.trigger('click');

        expect(document.body.classList.contains('fhamburgerswitch-on')).toBe(true);

        await wrapper.trigger('click');

        expect(document.body.classList.contains('fhamburgerswitch-on')).toBe(false);
    });
});
