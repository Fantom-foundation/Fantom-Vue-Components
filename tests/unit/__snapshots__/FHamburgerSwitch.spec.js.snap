// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FHamburgerSwitch should render correctly 1`] = `
<div
  class="fhamburgerswitch fhamburgerswitch-thickness1"
  role="switch"
>
  <div
    class="cont"
  >
    <div
      class="l1"
    />
     
    <div
      class="l2"
    />
     
    <div
      class="l3"
    />
  </div>
</div>
`;
